8354954: Typed static memory for late initialization of static class members in Hotspot

jdksjolen opened the pull request:
Hi,

This PR introduces a `StableValue<T>` which is sized and aligned identically to a `T`, with the difference that a `StableValue<T>` needs to be explicitly instantiated.

Dynamic static initalization in C++ leads to unpredictable bugs as there is no defined order in which objects will be initialized, to the degree that 'static initialization fiasco' is a term used. In the code I've worked on in Hotspot we resolve this by having an initialization function, and instead of having static members of `T` we have `T*` instead and use `malloc` in order to gain the memory for the objects. This is workable, but is unnecessary.

That's why I'd like to have `StableValue<T>`. It let's you avoid the whole `malloc` thing, and we overload `->` to make it behave as if it is actually a `T`. We add in a simple checker in debug mode that checks whether the memory has been initialized before using it.

In the code I've switched two members to be of `StableValue` instead. One is the malloc case above, the second (MemBaseline) is one where I got a bug while developing. The bug occurred because I  changed the initializer of `MemBaseline` without knowing that it was dynamic-static-allocated, and the exact change I made caused weird crashes (because of initialization order issues).

This solution is quite practical to me, but I wanted to know what others think.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8354954](https://bugs.openjdk.org/browse/JDK-8354954): Typed static memory for late initialization of static class members in Hotspot (**Enhancement** - P4)


### Reviewers
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24689/head:pull/24689` \
`$ git checkout pull/24689`

Update a local copy of the PR: \
`$ git checkout pull/24689` \
`$ git pull https://git.openjdk.org/jdk.git pull/24689/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24689`

View PR using the GUI difftool: \
`$ git pr show -t 24689`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24689.diff">https://git.openjdk.org/jdk/pull/24689.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24689#issuecomment-2813199150)
</details>

JornVernee commented on src/hotspot/share/nmt/memoryFileTracker.cpp:
```diff
@@ -126,9 +126,7 @@ const GrowableArrayCHeap<MemoryFileTracker::MemoryFile*, mtNMT>& MemoryFileTrack
-  _tracker = static_cast<MemoryFileTracker*>(os::malloc(sizeof(MemoryFileTracker), mtNMT));
-  if (_tracker == nullptr) return false;
-  new (_tracker) MemoryFileTracker(tracking_level == NMT_TrackingLevel::NMT_detail);
+  new (_tracker.as()) MemoryFileTracker(tracking_level == NMT_TrackingLevel::NMT_detail);
```
Maybe you could add an `init` function that forwards the constructor arguments, with an extra check to see if the memory has already been initialized:

```
  template<typename...As>
  void init(As&&...args) {
    assert(is_death_pattern(), "StaticArea already initialized");
    new (as()) T(std::forward<As>(args)...);
  }
```

```suggestion
  _tracker.init(tracking_level == NMT_TrackingLevel::NMT_detail);
```

jdksjolen replied:
I think that's a good idea. Unfortunately, move semantics and Rvalue references are currently undecided in the style guide, so we can't write this exact code. We can still do

```c++
template<typename... As>
void init(As&... args) {
    assert(is_death_pattern(), "StaticArea already initialized");
    new (as()) T(args...);
}
```

Which is pretty good.

merykitty commented:
Your implementation has a lot of UB, I believe a conformant implementation should look like this:

    class StaticArea {
        union {
            T _data;
        };
        DEBUG_ONLY(bool _present);

        StaticArea() : _present(false) {}

        T* as() {
            assert(_present, "");
            return &_data;
        }

        template <class Ts...>
        void init(Ts&... args) {
            assert(!_present, "StaticArea already initialized");
            new (&_data) T(args...);
        }
    };

merykitty commented:
P.S: I saw where your confusion comes from now :) You need an anonymous union member, you also need a destructor for `S`, like this:

    struct S {
        union {
            F _f;
        }

        ~S() {}
    }

jdksjolen commented on src/hotspot/share/utilities/staticArea.hpp:
```diff
@@ -0,0 +1,66 @@
+#endif
+
+  T* as() {
+    return reinterpret_cast<T*>(this);
```
Then I don't see how we can do this, as we really want to avoid having to store a pointer.

jdksjolen commented on src/hotspot/share/utilities/staticArea.hpp:
```diff
@@ -0,0 +1,66 @@
+  }
+  T* operator->() {
+    assert(!is_death_pattern(), "Potential access to uninitialized memory");
+    return as();
```
We don't compile with strict aliasing enabled, so I think this is fine, even if it goes againt the spec.

jdksjolen commented:
> P.S: I saw where your confusion comes from now :) You need an anonymous union member, you also need a destructor for `S`, like this:
> 
> ```
> struct S {
>     union {
>         F _f;
>     }
> 
>     ~S() {}
> }
> ```
> 
> https://godbolt.org/z/xdTKTjWE7

Yikes, I can't believe that it matters whether the union is a named type or not :-(. The `death_pattern` stuff is UB under strict aliasing and with a `T` with less-than-4 byte alignment, I agree with that. You might be correct with regards to `std::launder`, I guess it's a pointer provenance issue.

I'm not against fixing UB per se, so I prefer the union approach here.

For those following along, this is an example of `std::launder`:

https://miyuki.github.io/2016/10/21/std-launder.html

jdksjolen commented:
@merykitty,

Is this the correct implementation??

```c++
#include <new>

// A memory area with adequate size and alignment for storage of a T.
template<typename T>
class alignas(alignof(T)) StaticArea {
    union {
        T _t;
    };

public:
  StaticArea() {

  }
  ~StaticArea() {}

  T* operator->() {
        return as();
  }

    T* as() {
        return std::launder(reinterpret_cast<T*>(this));
    }

  template <class... Ts>
  void init(Ts&... args) {
    new (as()) T(args...);
  }
};
```

jdksjolen commented:
Alright, after reading up on this: This can only be done in legal C++17, in C++14 we have to save the old pointer in order to be non-UB.

merykitty commented:
You can access return the pointer to `_t` directly without having to `reinterpret_cast` from `this`. Also, a union must have the size and alignment to contains a `T` so you don't need `alignas`, either.

JornVernee commented:
Not that strict aliasing is turned off for hotspot (see `make/autoconf/flags-cflags.m4`). That said, the latest proposed version looks very good (with the suggestion to change the impl of `as()` to `return &_t;`).

jdksjolen commented:
Here's a new version. Do we even need the laundry if we take a pointer to `_t` instead? I don't think so.

TheShermanTanker commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,72 @@
+#include "globalDefinitions.hpp"
+
+// Pre-C++17 hack: Use  __builtin_launder when available.
+// Replace this with std::launder when we've upgraded.
```
We may have to revisit the Style Guide if we want to use the <new> header if and when we get to C\+\+17

jdksjolen commented:
I renamed the `.as()` function to `.ptr()`

jdksjolen commented:
Oops, I broke the build. I should also add some unit tests to this.

I'm off on a long Easter holiday, so plenty of time to get your opinions in if something else is missing.

tstuefe commented:
> ... however when working with certain code you may want to allocate before NMT is initialized. This leads to you having to use raw malloc. This is probably something you find out after having used os::malloc and already had a strange crash when building your code.

? No, this should not happen. You can, of course, use `os::malloc` before NMT is initialized; that is why NMT preinit mode exists. If you get strange crashes (do you have JBS issues for them?), those are bugs and should be fixed.

If this is the only motivation for this PR, we may be able to save a lot of time.

jdksjolen commented:
> > ... however when working with certain code you may want to allocate before NMT is initialized. This leads to you having to use raw malloc. This is probably something you find out after having used os::malloc and already had a strange crash when building your code.
> 
> ? No, this should not happen. You can, of course, use `os::malloc` before NMT is initialized; that is why NMT preinit mode exists. If you get strange crashes (do you have JBS issues for them?), those are bugs and should be fixed.
> 
> If this is the only motivation for this PR, we may be able to save a lot of time.

You're right, I managed to confuse this with another bug I had, where I mmapped before the VMT was intialized. Sloppy of me.

Happy Easter, let's talk more later.

merykitty commented:
> Saving this in memory as another reason for transitioning to C++17

Please, I really look forward to `if constexpr` :)

dholmes-ora commented:
@jdksjolen  Can you please give some concrete examples of existing things in the codebase that have this initialization problem and which would be candidates for this `StableValue`?

Also I think the name will have to change as it may be confused with "stable values" as they relate to Java code and compiler optimizations.

Happy Easter. :)

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,72 @@
+class StableValue {
+  union {
+    T _t;
+  };
```
All of this stuff about laundering and this union trick could go away if we're willing to add a pointer
to the StableValue class.  Something like this (not tested):
```
template<typename T>
class StableValue {
  alignas(T) char _data[sizeof(T)];
  T* _p;
public:
  StableValue() : _p(nullptr) {}
  ~StableValue() = default;
  NONCOPYABLE(StableValue);
  T* ptr() {
    assert(_p != nullptr, "uninitialized");
    return _p;
  }
  T* operator->() { return ptr(); }
  T& operator*() { return *ptr(); }
  template<typename... Args>
  void initialize(Args&&... args) {
    assert(_p == nullptr, "already initialized");
    _p = ::new (_data) T(std::forward<Args>(args)...);
  }
};
```
Regarding the destructor, quoting from the style guide: "HotSpot doesn't generally try to cleanup
on exit, and running destructors at exit can also lead to problems."

merykitty replied:
This would induce an additional indirection on every access, though.

kimbarrett replied:
That additional indirection is what we already have with the current idiom.  I'm looking for a way to be
safe while removing the "do we need to do laundry" question.  But I'm currently thinking we don't need
to do laundry, though I'm still working through the details of the rationale, and still might end up deciding
I'm wrong.

kimbarrett replied:
I think something this works, and that no laundering is required.
```
template<typename T>
class StableValue {
  union { T _obj; };
  DEBUG_ONLY(bool _initialized;)
public:
  StableValue() DEBUG_ONLY(: _initialized(false)) {}
  ~StableValue() {}
  NONCOPYABLE(StableValue);
  T* ptr() {
    assert(_initialized, "uninitialized");
    return &_obj;
  }
  T* operator->() { return ptr(); }
  T& operator*() { return *ptr(); }
  template<typename... Args>
  void initialize(Args&&... args) {
    assert(!_initialized, "already initialized");
    using NCVP = std::add_pointer_t<std::remove_cv_t<T>>;
    ::new (const_cast<NCVP>(_obj)) T(std::forward<Args>(args)...);
  }
};
```
[Note that I've made a few other improvements, such as supporting a cv-qualified
type and being non-copyable.  Obviously there's lots of missing comments...]

The rationale for not needing laundering is as follows.

std::launder is introduced to work around some restrictions involving reuse of
storage for a different object.  Specifically, C++17 6.8.8 (Object lifetime)
says

--- begin quote ---

If, after the lifetime of an object has ended and before the storage which the
object occupied is reused or released, a new object is created at the storage
location which the original object occupied, a pointer that pointed to the
original object, a reference that referred to the original object, or the name
of the original object will automatically refer to the new object and, once
the lifetime of the new object has started, can be used to manipulate the new
object, if:

[... list of requirements ...]

[Note: Note: If these conditions are not met, a pointer to the new object can
be obtained from a pointer that represents the address of its storage by
calling std::launder. -- end note.]

--- end quote ---

Other than it's definition and an example, that's the only occurrence of
"launder" in C++17.

But if we're careful, that section doesn't apply at all.  By using the
uninitialized anonymous union trick (or a similarly uninitialized and
appropriately aligned and sized char array), we have storage but not an object
whose lifetime has ended.  There wasn't a prior object in that storage.

Instead we're dealing with storage either before or after the object's
lifetime has started. There are restrictions on what one can do with a pointer
to storage before the lifetime of the object starts, but we're not violating
any of those.  The object's lifetime starts with the placement-new.

merykitty replied:
I believe you are right, we do not need launder here. From [[class.union]/6](https://eel.is/c++draft/class.union#6):

> [Note [5](https://eel.is/c++draft/class.union#general-note-5): In cases where the above rule does not apply, the active member of a union can only be changed by the use of a placement [new-expression](https://eel.is/c++draft/expr.new#nt:new-expression)[.](https://eel.is/c++draft/class.union#general-6.sentence-1) — end note]
[Example [3](https://eel.is/c++draft/class.union#general-example-3): Consider an object u of a union type U having non-static data members m of type M and n of type N[.](https://eel.is/c++draft/class.union#general-6.sentence-2) If M has a non-trivial destructor and N has a non-trivial constructor (for instance, if they declare or inherit virtual functions), the active member of u can be safely switched from m to n using the destructor and placement [new-expression](https://eel.is/c++draft/expr.new#nt:new-expression) as follows:
u.m.~M();
new (&u.n) N;
— end example]

We can see that `::new(_&obj) T()` is considered an act to make `_obj` active, not to create a new `T`, the `_obj` itself has been created before at this location, it is just that it has not been initialized and its lifetime has not started. As a result, `_obj` can be used to refer to this object without launder.

merykitty replied:
To be perfectly clear, we can just remove launder from the patch, and do not need to do this:

    template<typename... Args>
    void initialize(Args&&... args) {
      assert(!_initialized, "already initialized");
      using NCVP = std::add_pointer_t<std::remove_cv_t<T>>;
      ::new (const_cast<NCVP>(_obj)) T(std::forward<Args>(args)...);
    }

Just do it normal:

    template<typename... Args>
    void initialize(Args&&... args) {
      assert(!_initialized, "already initialized");
      ::new (&_obj) T(std::forward<Args>(args)...);
    }

jdksjolen replied:
Good, I've removed the launder. I see your point re: const-ness.

merykitty commented:
Another benefit of this patch is that it reduces 1 indirection accessing `T` compared to a pointer, which I think is worth it.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,72 @@
+  ~StableValue() {
+  }
+
+  T* ptr() {
```
More naming bike shedding.  This is kind of a smart pointer type.  The "usual"
name for this operation on smart pointers seems to be `get`.

jdksjolen commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,72 @@
+  template<typename... Ts>
+  void initialize(Ts&... args) {
+    DEBUG_ONLY(_initialized = true);
+    new (ptr()) T(args...);
```
I agree, but can we do this in a separate PR then? Getting in a style guide change via this PR wouldn't be very nice.

kimbarrett replied:
Personally, I'm okay with this using perfect forwarding now.

jdksjolen commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,72 @@
+// A memory area with adequate size and alignment for storage of a T.
+// May be initialized exactly once.
+template<typename T>
+class StableValue {
```
I prefer fully spelled out words. David's comment re stable values in Java rings true, it's not a good name. I'll wait with renaming, see if something else is suggested. 'Deferred' and 'Delayed' are good prefixes. Could *just* be 'Delayed<T>'.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+
+public:
+  StableValue() {
+    DEBUG_ONLY(_initialized = false);
```
Prefer ctor-initializer initialization to assignment in the body.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+
+  T* operator->() {
+    assert(_initialized, "must be initialized before access");
+    return ptr();
```
Just call `ptr()`, which does the assert.

kimbarrett replied:
I think this class should also provide `operator*()`.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+
+  T* ptr() {
+    assert(_initialized, "must be initialized before access");
+    return &this->_t;
```
Why `&this->_t` rather than `&_t`?

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+
+  ~StableValue() {
+  }
+
```
This class should be noncopyable.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+  void initialize(Ts&... args) {
+    DEBUG_ONLY(_initialized = true);
+    // If T has const and volatile, get rid of them and tack on a pointer *.
+    using NCVP = std::add_pointer_t<std::remove_cv_t<T>>;
```
I think the comment is grammatically confusing as written.  I also think the comment doesn't add any
information, and we'd be better off just removing that comment.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+    DEBUG_ONLY(_initialized = false);
+  }
+
+  ~StableValue() {
```
Comment the ctor/dtor as intentionally not initializing `_t` and not destroying it either.

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,68 @@
+    // If T has const and volatile, get rid of them and tack on a pointer *.
+    using NCVP = std::add_pointer_t<std::remove_cv_t<T>>;
+    // Make _t into NCVP temporarily so that we can placement-new it.
+    new (const_cast<NCVP>(ptr())) T(args...);
```
s/new/::new/ to avoid looking up the allocator in the scope of `T`.  All of the uses in this PR are
for "value classes" rather than (say) `CHeapObj`, so don't provide `operator new` &etc.

kimbarrett commented on src/hotspot/share/nmt/memTracker.hpp:
```diff
@@ -261,7 +261,7 @@ class MemTracker : AllStatic {
   // Stored baseline
   static inline MemBaseline& get_baseline() {
-    return _baseline;
+    return *_baseline.ptr();
```
I think StableValue should provide `operator*()` and it should be used here.

jdksjolen commented:
@dholmes-ora , @tstuefe 

I hope you had a good Easter :).

>@jdksjolen Can you please give some concrete examples of existing things in the codebase that have this initialization problem and which would be candidates for this StableValue?

To be clear, no issues exist today because we circumvent the issue using malloc. I'm suggesting an alternative way of circumventing the problem, while allowing us to more clearly express the intent of our code. If we have code like this:

```c++
class Foo {
  static Bar* _bar;
  // ...
};
```

Here, `_bar` may be reassigned, it may be null and it has an unknown lifetime. When I'm reading the code, I want to ensure that I understand it, so I may look at how often and when it is assigned and so on.

If we have this, instead:

```c++
class Foo {
  static StableValue<Bar> _bar;
};
```

Then we know that `_bar` will be assigned exactly once, that it will be uninitialized prior to the point that it's initialized, it cannot be null,  and that it will never be deallocated.

I think that conveys a lot more about the specific usage of the value, just from reading that line. **This PR  doesn't aim for further correctness**, it aims for developer ergonomics. This class would make my life easier and less tedious when writing and reading code, and I'm hoping that others agree that it would help them as well.

In the PR I give two concrete places where I switch to using StableValue, but there are of course more.

See `ZInitialize::initialize` and `ZDriver::initialize`. There's a strict initialization order, as objects need to materialize in the order that they depend on each other. The `ZLock` in `ZDriver::initialize` is `malloc`ed, why? Probably because they didn't want to take the risk of having a static initialization fiasco.

jdksjolen commented:
@kimbarrett, I'll be waiting with fixing your comments (they're good comments!). I'd like to have some stronger consensus that we want to have this integrated before committing more time :).

jdksjolen commented:
@kimbarrett,

It's been 5 days and we've received no more comments. I've applied all of your comments, except the `get()` rename. To me, both `ptr` and `get` is fine. We can take the rename just to make it more obviously pronuncable.

I haven't run any testing except building on the current HEAD, so we'll wait for GHA.

We need a better name than `StableValue<T>`. We could call it `Deferred<T>`, or `DeferredInitialization<T>`. I think that `Deferred<T>` is better, it's a bit shorter while conveying *some* meaning. You can very easily jump2def to see what it says in the docs, and then probably remember it.

Regarding tooling: I use `clangd` as my LSP server, so clang provides my IDE capabilities. It has no problem with figuring out the T and using it after a `get()` or `->`.

JornVernee commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,76 @@
+
+  template<typename... Ts>
+  void initialize(Ts&... args) {
+    DEBUG_ONLY(_initialized = true);
```
Shouldn't we also check here that `_initialized == false`?

kimbarrett replied:
Yes, it should.  My (untested) proto earlier had the other, more serious, bug of forgetting to
assign `_initialized`!

jdksjolen replied:
Yup

kimbarrett commented on src/hotspot/share/utilities/stableValue.hpp:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  T& operator*() {
+    return *get();
```
Why not just `return _t;`?

jdksjolen replied:
To get the initialization assert.

kimbarrett replied:
Doh!  Of course!

kimbarrett commented:
> @kimbarrett,
> 
> I've applied all of your comments, except the `get()` rename. To me, both `ptr` and `get` is fine. We can take the rename just to make it more obviously pronuncable.

Looks like the rename to `get()` happened after all?

> We need a better name than `StableValue<T>`. We could call it `Deferred<T>`, or `DeferredInitialization<T>`. I think that `Deferred<T>` is better, it's a bit shorter while conveying _some_ meaning. You can very easily jump2def to see what it says in the docs, and then probably remember it.

`Deferred<T>` seems fine to me.

> Regarding tooling: I use `clangd` as my LSP server, so clang provides my IDE capabilities. It has no problem with figuring out the T and using it after a `get()` or `->`.

I'm not sure what this is in response to?  Maybe to a comment from someone else?

jdksjolen commented:
> > @kimbarrett,
> > I've applied all of your comments, except the `get()` rename. To me, both `ptr` and `get` is fine. We can take the rename just to make it more obviously pronuncable.
> 
> Looks like the rename to `get()` happened after all?
> 

Yeah, I thought about it and applied the change after posting the comment :). Or maybe I just forgot to update the comment!

> > Regarding tooling: I use `clangd` as my LSP server, so clang provides my IDE capabilities. It has no problem with figuring out the T and using it after a `get()` or `->`.
> 
> I'm not sure what this is in response to? Maybe to a comment from someone else?

Sorry, this wasn't in response to anything you said. I wanted this to be part of the record of the PR, as I know folks are generally wary of how templates affect their IDEs capability to analyze the code.

jdksjolen commented:
I've added an assert that any T should only have a trivial destructor, as any user-defined destructor will not be called.

jdksjolen commented:
> I've added an assert that any T should only have a trivial destructor, as any user-defined destructor will not be called.

Aaand, I broke the build :). I guess that that might be too much of a stretch, as it makes it impossible to use certain types in certain situations. I reverted the change and deleted the assert.

jdksjolen commented:
Test failure in Shenandoah, unrelated timeout.

kimbarrett commented on src/hotspot/share/utilities/deferred.hpp:
```diff
@@ -0,0 +1,76 @@
+  Deferred()
+  DEBUG_ONLY(: _initialized(false)) {
+    // Do not construct value, on purpose.
+  }
```
I found this placement of the debug-only ctor-init eye-catchingly confusing, and would have myself
placed it on the line above.  But I won't insist if others are okay with it as is.

kimbarrett reviewed (approved):
Looks good.

kimbarrett commented:
> I guess we decided not to go with perfect forwarding in this PR?

I overlooked that.  I'd prefer that perfect forwarding was used here.  But I won't block if it isn't.

jdksjolen commented:
> > I guess we decided not to go with perfect forwarding in this PR?
> 
> I overlooked that. I'd prefer that perfect forwarding was used here. But I won't block if it isn't.

Right, I want perfect forwarding but that's more up to style guide update rather than this PR.

jdksjolen commented:
Thank you, all!

/integrate
